marksman: init at 2022-12-28

stasjok opened the pull request:
###### Description of changes

Language server for Markdown: https://github.com/artempyanykh/marksman.

Closes #184873.

I plan to use it myself and thought that I can try to add it to nixpkgs. Never used dotnet, but based on my basic testing, the build is successful and the app is working (tested on linux).

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

stasjok committed:
marksman: init at 2022-12-28

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1225

TLATER commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,50 @@
+, dotnetCorePackages
+}:
+
+buildDotnetModule rec {
```
Nit: If possible, it would really be better to use the new [recursive attributes feature of `mkDerivation`](https://nixos.org/manual/nixpkgs/stable/#idm140737319947216), but I don't know if `buildDotnetModule` supports that yet.

stasjok replied:
> but I don't know if `buildDotnetModule` supports that yet.

I don't think it does.

SuperSandro2000 replied:
Why would we want to use that here? Just to substitute version in the fetcher? The fetcher is FOD anyway, so it would only work when taking extra care to overwrite the fetcher which is not documented anywhere. More the opposite is that src is always overwritten in all examples. Also people tend to use e.g. pname all over the place which would catastrophically fail with finalAttrs.

TLATER commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,50 @@
+    '';
+    homepage = "https://github.com/artempyanykh/marksman";
+    license = licenses.mit;
+    maintainers = with maintainers; [ ];
```
Are you planning to contribute this package without a maintainer? Or are you simply unsure as to whether it's ok to add yourself to the contributor list?

SuperSandro2000 replied:
meta.maintainers must be set otherwise we cannot accept the PR.

TLATER commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,50 @@
+  doCheck = true;
+  testProjectFile = "Tests/Tests.fsproj";
+
+  nugetDeps = ./deps.nix;
```
Since this is a relatively complex update process, it would be really nice if there were an [update script](https://nixos.org/manual/nixpkgs/stable/#var-passthru-updateScript) for this one, especially if you don't intend to maintain it yourself.

I've never used dotnet either, but I imagine it wouldn't be too hard to do the nuget fetching in a script?

TLATER commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,50 @@
+      references that enable Zettelkasten-like note taking.
+    '';
+    homepage = "https://github.com/artempyanykh/marksman";
+    license = licenses.mit;
```
MIT requires:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

Currently this isn't copied into the package:

```console
tlater ~/result $ ls
bin  lib
tlater ~/result $ fd LICENSE .
```

Arguably a problem with their build scripts, I suppose, but let's copy it in, given their license asks for it? Something like

```nix
postInstall = ''
  mkdir -p $out/share/doc
  install -m 0644 LICENSE $out/share/doc/
'';
```

This isn't always upheld in nixpkgs, but it's a tiny amount of effort, so...

SuperSandro2000 replied:
> This isn't always upheld in nixpkgs

Yes, because it creates conflicts between packages when installed globally.

TLATER reviewed (approved):
Looks good to me and works fine too, at a quick use. I left some comments, but it looks mostly fine to me when compared to existing nixpkgs packages :)

stasjok commented:
I've added updateScript and a very simple test.

TLATER commented:
Final note, the commits need to be squashed, I'll add it to the reviewed-prs post once that's done.

stasjok commented:
> Final note, the commits need to be squashed, I'll add it to the reviewed-prs post once that's done.

Done.

TLATER commented:
Result of `nixpkgs-review pr 193682` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>marksman</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/675

SuperSandro2000 commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,59 @@
+
+buildDotnetModule rec {
+  pname = "marksman";
+  version = "2022-10-30";
```
```suggestion
  version = "unstable-2022-10-30";
```

stasjok replied:
I didn't manage to build it with this version.

1. Version starting with a string is not allowed in dotnet.
1. Looks like dotnet is using `Version` or `version` env to lookup a version, or a tag from project file. But in a build phase nix overrides even that:
https://github.com/NixOS/nixpkgs/blob/b457130e8a21608675ddf12c7d85227b22a27112/pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-build-hook.sh#L23-L25

And to be honest I don't like it much. I know about the rule:

> If a package is not a release but a commit from a repository, then the version attribute must be the date of that (fetched) commit. The date must be in `unstable-YYYY-MM-DD` format.

But it's not a random commit, it's a tagged release on github which happens to be a date. But yes, it looks like there is no traditional versioning for this project yet.

Any suggestions?

TLATER replied:
`dotnetBuildFlags` seem to come after that version flag, does dotnet allow overriding arguments so you can hard-code your own version flag to override the one the build phase sets?

stasjok replied:
I already tried that without luck. I tested it again and my problem was with configure phase. With `dotnetFlags` it works. But it somewhat complicates things, especially with update script.

stasjok replied:
OK. Even if I don't like it, since no one is answering, I've added `unstable-` prefix. I hope it can be merged now.

SuperSandro2000 commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,59 @@
+    sha256 = "sha256-Ko/dHV0mxRYukr9vdQD7HSDOYdqfc/DcuU8XPCYkSWA=";
+  };
+
+  patches = [ ./patches/0001-revert-embed-git-commit-into-the-version.patch ];
```
```suggestion
  patches = [ ./0001-revert-embed-git-commit-into-the-version.patch ];
```

SuperSandro2000 commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,59 @@
+  dotnet-sdk = dotnetCorePackages.sdk_6_0;
+  dotnet-runtime = dotnetCorePackages.runtime_6_0;
+
+  postInstall = ''
+    install -m 644 -D -t $out/share/doc LICENSE
+  '';
+
```
This will conflict with any other program that puts a file there and create needless conflicts.

```suggestion
```

TLATER replied:
How about

```
postInstall = ''
    install -m 644 -Dt $out/share/doc/marksman/ LICENSE
'';
```

instead then? At least on my system a few packages seem to already be placing license files along with documentation there, and this matches what other distros do.

stasjok replied:
Yes, if we want to put a LICENSE, it'll have to be `$out/share/doc/marksman`. I didn't notice that it's in common dir, my bad.

SuperSandro2000 commented on pkgs/development/tools/marksman/update.sh:
```diff
@@ -0,0 +1,18 @@
+shopt -s inherit_errexit
+
+version=$(curl -s https://api.github.com/repos/artempyanykh/marksman/releases/latest | jq -r .tag_name)
+old_version=${UPDATE_NIX_OLD_VERSION:-$(nix-instantiate --eval -A marksman.version | jq -r)}
```
```suggestion
old_version=$(nix-instantiate --eval -A marksman.version | jq -r)
```

SuperSandro2000 commented on pkgs/development/tools/marksman/update.sh:
```diff
@@ -0,0 +1,18 @@
+#shellcheck shell=bash
+
+set -eu -o pipefail
+shopt -s inherit_errexit
```
```suggestion
```
We are not setting this in many update scripts so I am a bit reluctant to do it here.

SuperSandro2000 commented on pkgs/development/tools/marksman/update.sh:
```diff
@@ -0,0 +1,18 @@
+set -eu -o pipefail
+shopt -s inherit_errexit
+
+version=$(curl -s https://api.github.com/repos/artempyanykh/marksman/releases/latest | jq -r .tag_name)
```
This should optionally take a GITHUB_TOKEN

```suggestion
version=$(curl -s ${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""} https://api.github.com/repos/artempyanykh/marksman/releases/latest | jq -r .tag_name)
```

stasjok replied:
Just a note. I noticed this `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` in some places in nixpkgs, but I think it's just wrong. This substitution expands to one word: `-u ":$GITHUB_TOKEN"`. Correct substitution `${GITHUB_TOKEN:+-u ":$GITHUB_TOKEN"}` expands to two words: `-u` and `:$GITHUB_TOKEN`.

You can test it:

```bash
stas@server2:~$ a="aa bb"
stas@server2:~$ set -x
stas@server2:~$ : ${a:+"-u \":$a\""}
+ : '-u ":aa bb"'
stas@server2:~$ : ${a:+-u ":$a"}
+ : -u ':aa bb'
```

In essence with `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` curl will use username ` "` (space and quote) and password `$GITHUB_TOKEN"` (with quote at the end). Token will be invalid.

SuperSandro2000 replied:
Can you fix this up, please?

``GITHUB_API_TOKEN`` is just another alias for ``GITHUB_TOKEN`` and probably the 5th I have in my configuration.

SuperSandro2000 replied:
I just did a double check and it seems to expand correct. I think we are missing an outer quote, so it is resplit.

SuperSandro2000 replied:
> You still think that `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` is correct?

No, but it works probably due to resplicing.

SuperSandro2000 replied:
#204219

stasjok commented:
I believe I addressed most of the conversations.

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -12979,6 +12979,10 @@
   };
+  stasjok = {
+    name = "Stanislav Asunkin";
+    email = "nixpkgs@stasjok.ru";
```
Please add github and githubId.

SuperSandro2000 commented on pkgs/development/tools/marksman/default.nix:
```diff
@@ -0,0 +1,63 @@
+
+buildDotnetModule rec {
+  pname = "marksman";
+  releaseTag = "2022-10-30";
+  version = "unstable-${releaseTag}";
```
```suggestion
  version = "2022-10-30";
```
there was a confusion here. I thought this was an unstable tag.

michaeladler commented:
Nice work (once upton a time I tried to package as well but got too frustrated with the whole dotnet packaging experience)! I've cherry-picked your commit and the LSP works fine in neovim.

edrex commented:
There's a new release https://github.com/artempyanykh/marksman/releases/tag/2022-11-25
Better to bump before merge?

edrex commented:
New major feature release: https://github.com/artempyanykh/marksman/releases/tag/2022-12-19

avdv commented:
This seems good to go. Anything left to do here?

stasjok commented:
> New major feature release: https://github.com/artempyanykh/marksman/releases/tag/2022-12-19

I just pushed the update.

> This seems good to go. Anything left to do here?

No. It's ready.

avdv commented:
@stasjok thanks for the update! :+1: 

Could you take another look @SuperSandro2000 ? Thanks!

stasjok commented:
1. Updated to **2022-12-23**.
2. Dropped a patch that prevented running git during build process. Thanks to artempyanykh/marksman#125.

Now the version string reported by `marksman --version` is `1.0.0-2022-12-23` (upstream binaries have version like `1.0.0-662682b`). `1.0.0` is still hard-coded here: https://github.com/artempyanykh/marksman/blob/662682b9d5fc163ed4be3ce0dba54febd6258bf3/Marksman/Marksman.fsproj#L7

@SuperSandro2000, what do you think if we change a version of the package to `1.0.0-<git tag>` too? I don't think that `1.0.0` has any meaning upstream, it's like a placeholder for version (at least for now).

pacien reviewed (approved):
Imported and tested on 22.11.
It is automatically detected and used by the Helix editor.
Everything seems to be working fine.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/784

SuperSandro2000 commented:
If a software tags releases we use those tags as the version number and strip prefixes like v, r or release.
